activemodel/lib/active_model/secure_password.rb

```diff
@@ -155,7 +155,7 @@ def has_secure_password(attribute = :password, validations: true, reset_token: t
             end
           end
 
-          validates_confirmation_of attribute, allow_blank: true
+          validates_confirmation_of attribute, allow_nil: true
```

Comment:
This change assumes that, given the current implementation of the writer method `password=`, the corresponding reader method `password` can only actually return `nil` or a string that is not `empty?` (whitespace-only or not). Therefore, by skipping the confirmation validation only when `password` is `nil`, we get to validate every non-empty string, which is what we want.

Reply:
To my knowledge, an empty form `<input>` should not result in a `nil` value, but nonetheless, `nil` as a `password` value (e.g. assigned by other means) is explicitly supported, and has the effect of setting `password_digest` to `nil` as well, see https://github.com/rails/rails/blob/main/activemodel/lib/active_model/secure_password.rb#L187. We want to skip confirmation in that case alone.
Plus, I tried to remove the `allow_` condition altogether, but that resulted in test failures.
Let me know if I answered your question!

Reply:
Makes sense to me. If there was a confirmation field in the form, you'll get `""`, if the field is being changed from another context, it just won't be there at all (`nil`).